### PR TITLE
[READY] Smart Cables

### PR DESCRIPTION
**Pull request by actioninja**
## About The Pull Request
image

Cables now autoconnect on cardinal directions. All cable placement has been completely stripped out and replaced with simple single cable per tile logic.
Low effort demo: https://www.youtube.com/watch?v=fXp8s6ORWbA
Yes I am aware that cutting it is not dropping wire, that version was bugged.
Cables no longer need a knot specifically placed to link to power objects. The sprite is automatically changed to represent this.
The only exception to this logic is that on smes units, due to the terminal being next to the output, they will not link there.
On a technical powernet side, this is the same as old cables once actually placed. They still use the existing powernet system, just the placement and connection works differently.

Old cables have been turned into "pipe cleaners" for wire art purposes. they work just like the old ones, just missing all the powernet functionality, and also you can put them on top of the floor.

## Why It's Good For The Game

How obnoxious cables were to both map and work with in game has been something that has annoyed me for a really long time now.
This is both easier for new players to learn, and easier for experienced players to work with.
Along with making in game much more intuitive and easier, it makes mapping much easier as well. Mad lad wjohn was able to rip out all the mapping conversion in one day because of how much faster it is to work with. Here's how the screenshot above looks in DM (on an older version that had red cable):
image

## Changelog
:cl: actioninja and wjohn
add: Cables have been completely reworked. Simple per tile connection logic, automatically connects to things above it. Think minecraft redstone.
add: Old cables have been kept as pipe cleaner. They are non-functional in terms of power, but otherwise have the same connection logic. Also can go on top of tiles.
remove: mech cable layer has been removed because it was terrible shitcode nobody used
tweak: (sort of balance) cable stack sized has been reduced to 15.
/:cl:

### CODE REVIEW

**Comment by actioninja** [comment]
map merge conflicts oh god oh fuck

**Comment by Fikou** [comment]
there needs to be a new mechanic to replace wire art if this becomes a thing

**Comment by redmoogle** [comment]
Don't forget to take the wires out of EVERY files

**Comment by Nickkoranda** [comment]
@Fikou read the PR description
>Old cables have been turned into "pipe cleaners" for wire art purposes. they work just like the old ones, just missing all the powernet functionality, and also you can put them on top of the floor.

**Comment by AffectedArc07** [comment]
RIP Wireart

**Comment by 81Denton** [comment]
@AffectedArc07 read the PR description
>Old cables have been turned into "pipe cleaners" for wire art purposes. they work just like the old ones, just missing all the powernet functionality, and also you can put them on top of the floor.

**Comment by AffectedArc07** [comment]
You saw nothing

**Comment by actioninja** [comment]
I know there's a merge conflict in box in that commit, it's a spicy one.

**Comment by Barhandar** [comment]
Pipe cleaners are for smoking pipes, you're now obligated to add the interaction.

This change is rather literally dumb in that it makes wiring require no thinking whatsoever. Also the "grilles being powered if placed on any wire" WILL get you roasted several times and you'll have to change it to require at least as much effort (that is, insuls or RCL + time to actually do it) as knotting current wires + placing grille on top.

**Comment by Tlaltecuhtli** [comment]
put smart cables behind techwebs so bahrnahr is happy

**Comment by 81Denton** [comment]
>This change is rather literally dumb in that it makes wiring require no thinking whatsoever.
Wiring right now is about 1% thinking and 99% clicking tiny sprites. 
It's even worse for mappers who have to select and place a different cable direction every time it doesn't run in a straight line.

**Comment by Barhandar** [comment]
You hadn't needed to click tiny sprites in YEARS what are you on.

>It's even worse for mappers who have to select and place a different cable direction every time it doesn't run in a straight line.

That calls for a mapping helper/an opt-in mode for placing wires ingame, not for reworking entirely how wires are placed and connected, as pointed out by one of the maintainers.

**Comment by MrPerson** [comment]
This is literally the absolute best way to do wiring. The only use case we lose out on is the VERY rare N-to-W, E-to-S double turn situation where the two networks don't connect, but I think we'll live without that thanks.

Thanks for keeping wire art in too btw, that shit is hella fun.

**Comment by zxaber** [comment]
MrPerson, we could have a special object that connects like that. A sort of cross that consists of two insulated lines, perhaps created by using a wire on a rod.

I pitched the idea already, and I'm not sure if it was deemed necessary. Wjohn seemed happy enough to just map around the restriction. I suppose if there's ever a good enough reason to _want_ to do that, we could always add it later.

**Comment by AffectedArc07** [comment]
How are we gonna handle wire nodes next to each other like the SMES have

Doesn’t this mean we’re permanently hotwired. The SMES relies on two seperate nodes adjacent to eachother like —• •—

**Comment by kriskog** [comment]
This is great. Cables were a fucking mystery to me for so long when I started out, this makes them easy. The only thing that worries me is the lack of knotting, and interaction with grilles/machines. Imagine being able to make an electric grille everywhere there is a cable and how annoying that is going to be when some assistant gets hold of a stack of metal rods.

I don't know the solution to this, while also keeping it intuitive, but heres a few possibilities: 
-Make it only work through special conductive floor tiles. Made by metal and cable.
-Only works if there is no floor tile (dumb for every machine that needs a knot)
-Need to connect things manually by using a cable on them. So build a grille, use a cable on it to connect to cable underneath. Requires insuls to not get shocked, like making a node today.

**Comment by zxaber** [comment]
>How are we gonna handle wire nodes next to each other like the SMES have

From the PR body, it sounds like there's a special interaction of some sort to handle this;
"The only exception to this logic is that on smes units, due to the terminal being next to the output, they will not link there."

>-Only works if there is no floor tile (dumb for every machine that needs a knot)

This is the closest to how it already works- if you want to set up a random emitter somewhere, you have to open the floor up where the wire is, make a knot, and then plop your emitter on top. So the new system would simply remove the step about making the knot.

**Comment by Barhandar** [comment]
Also make the cables yellow for the same reason they were remapped to be yellow: so they're visible on waste pipes.

**Comment by Fikou** [comment]
> > How are we gonna handle wire nodes next to each other like the SMES have
> 
> From the PR body, it sounds like there's a special interaction of some sort to handle this;
> "The only exception to this logic is that on smes units, due to the terminal being next to the output, they will not link there."
> 
> > -Only works if there is no floor tile (dumb for every machine that needs a knot)
> 
> This is the closest to how it already works- if you want to set up a random emitter somewhere, you have to open the floor up where the wire is, make a knot, and then plop your emitter on top. So the new system would simply remove the step about making the knot.

but you can plop a floor tile back on the place with the knot and the emitter would still work

**Comment by zxaber** [comment]
Fair enough, but the only time I think I've seen this done is the emitter room on Delta. Most of the engine emitter rooms already have the floor removed, and if you're using emitters outside engineering (in Cargo to open boxes or somewhere else to fight a blob), you either don't bother or don't have time to replace the tile.

Anyway, it makes sense to me that you'd need the tile to be out of the way to connect machinery to the cables below it.

**Comment by MMMiracles** [comment]
`That calls for a mapping helper/an opt-in mode for placing wires ingame, not for reworking entirely how wires are placed and connected, as pointed out by one of the maintainers.`
unless you're suggesting making an external program dedicated to wiring, then no, this method is so much fucking better when it comes to mapping. if similar functionality could be applied to pipes as well when it comes to making long winding sections (maintenance pipes, hallways, ect) then i might experience nirvana.

my only complaint is the fact it considers any wire going over a grille as a shocked grille, instead of the knot method before. there's a lot more times where i'd rather a grille not be shocked (i.e maintenance) and making it really generalized made some parts of the map significantly more deadly than before.

**Comment by MrPerson** [comment]
It's VERY tempting to say that danger is a feature, not a defect, but @kriskog suggested some solutions. I like the third one. At least for grilles (maybe just grilles?).

**Comment by Barhandar** [comment]
>unless you're suggesting making an external program dedicated to wiring, then no, this method is so much fucking better when it comes to mapping.

Mapping helper would literally be this but for maps only you chumbis.

**Comment by AffectedArc07** [comment]
```
_maps/map_files/Deltastation/DeltaStation2.dmm:65094:error: undefined type: /obj/item/stack/cable_coil/red
_maps/map_files/Deltastation/DeltaStation2.dmm:65096:error: unknown type

_maps/map_files/MetaStation/MetaStation.dmm:76078:error: undefined type: /obj/structure/cable/yellow
_maps/map_files/MetaStation/MetaStation.dmm:76079:error: invalid variable name: icon_state
_maps/map_files/MetaStation/MetaStation.dmm:76090:error: unknown type

_maps/map_files/BoxStation/BoxStation.dmm:41690:error: undefined type: 
_maps/map_files/BoxStation/BoxStation.dmm:41690:error: expected )
```

**Comment by nicbn** [comment]
How is this going to handle the case where a cable is supposed to cross above another cable? For example: output of engine crosses station supply cable to go to the smeses

**Comment by WJohn** [comment]
By not having them cross, @nicbn. I was easily able to wire that so they don't cross on every map. Any instances where crossing wires happened in the first place to a nasty degree was due to poor wiring in the first place, like having wires cross over one another repeatedly in a tangled mess, or having the wires go through walls which you shouldn't do in the first place.

**Comment by MrPerson** [comment]
We can add a wire bridge if needed. A lot of games have them, ex Oxygen Not Included.

File: code/game/objects/items/RCL.dm
Comment (Flatgub) [inline code/game/objects/items/RCL.dm]: Probably wanna remove these underscores.

File: code/game/objects/items/RCL.dm
Comment (Flatgub) [inline code/game/objects/items/RCL.dm]: You'll need to either rename the icon states in radial.dmi to match `pipe_cleaner_0-1` and so on or leave this as `cable_[pipe_cleanersuffix]`

**Comment by ninjanomnom** [comment]
As said previously you can accomplish all this with a mapping helper for the mapping side and a placement mode for in game without losing to have crossing cable networks. I see no good reason to simplify the actual cables themselves like this

**Comment by nicbn** [comment]
I think it'd be cool if wires of different colors didn't connect unless the user explicitly wants it (by clicking the other cable or something like that), that way one could run multiple disconnected powernets parallel to each other

**Comment by PKPenguin321** [comment]
Could make it so you have to use a wire on a grille to make it a shocked grille. Would effectively preserve functionality

**Comment by Ziiro** [comment]
I am excited to run through maint and drop grills for people chasing me to run into

I mean this entirely unironically. It's fucking comical and great. I want it.

**Comment by actioninja** [comment]
@nicbn wires of different colors don't exist anymore because they're uneeded. I can't think of a reason why you'd want multiple powernets running parallel.

**Comment by actioninja** [comment]
@ninjanomnom name a single usecase where wires would 100% need to cross.

**Comment by ninjanomnom** [comment]
You're being obstinate and unimaginative if you think there's no usecases where wire crossing is useful. Multiple networks can start separate but expansion may eventually require their wires to cross. Yeah usually you could redo everything to make that not necessary or just use a single network, but making it impossible reduces the design space possible with cables.

Name a single usecase where you need different floor decals. Name a single usecase where you need more than one area in a shuttle. Name a single usecase where you need more than one color in a sprite. Not all things must be reduced to what is only strictly "necessary". The word itself being subjective.

Frankly, unless some other maintainer cares enough to argue otherwise, I'll be closing this pr after 24 hours.

**Comment by WJohn** [comment]
We thought about having different kinds of wires or wire bridges but really, 99% of the time the station doesn't use that level of complexity. This system simply speeds up the process for everybody.

**Comment by 81Denton** [comment]
@ninjanomnom Like WJohn said, the vast majority of all rounds only have a single powernet, not to mention that no map in the repo has two intersecting powernets.

In my opinion, the convenience of fast cabling outweighs the potential loss of complexicity. This also plays into speeding up repairs and construction without directly buffing any items or mechanics.

**Comment by WJohn** [comment]
We should also at least give this thing a testmerge to see how players like it? It seems kind of a waste to just close this down without even trying it.

**Comment by ninjanomnom** [comment]
Not different kinds of wires, you can accomplish everything you want to do here without any loss of actual functionality by just adding a placement mode for cables and a mapping helper for maps. The only trade being made here is to reduce the capability of cables for no reason when alternatives exist.

It doesn't matter if nothing uses it roundstart, this is just as much a tool in the toolbox for in game usage as it is for mappers.

**Comment by MrPerson** [comment]
We've had the option since literally forever and nobody's used it yet. I don't think anybody will. And quite frankly, the pain of actually laying cables is one of the big culprits why. Laying cables isn't fun, thus any feature that relies on players laying cables won't be fun, thus the "functionaility" that you're attempting to preserve is useless.

**Comment by actioninja** [comment]
@ninjanomnom the loss of functionality here is that you're adding even more complexity and obtuseness to wiring for no particular reason besides to support crossovers. When really there's very little reason to ever want to crossover, and to throw your own argument back at you, you're being unimaginative if you think crossovers are strictly necessary and the only way to keep things separate.
Modal UX is horrible for a lot of reasons, and another arbitrary and confusing mode setting and another stupid action button cluttering up the ui just to support a use case that is uneeded is exactly why this game's ux is such a clusterfuck.

**Comment by optimumtact** [comment]
you should add a merge wire that is multi colored (made by adding to bits of any colour wire together) and then don't join across wire types, so you can do really fancy wiring

**Comment by optimumtact** [comment]
anyway this is fine. Previous wire complexity was not complexity, it was clunkiness.

**Comment by optimumtact** [comment]
To be clear I will be merging this

**Comment by actioninja** [comment]
At the moment there's no wire colors since it wasn't actually needed for any base maps. I could roll colored wires with linking based on color into this PR but I think that's a better idea for a later date and to get this polished and refined first. I'll design it to be extensible so colored wired would be easier to add later.

**Comment by ninjanomnom** [comment]
@optimumtact you can accomplish the same goals as this pr without removing any complexity. This version just replaces the base systems with less capable systems for no particularly good reason. Keeping all the same actual functionality but just making cables automaticaly merge by default is completely possible.

**Comment by WJohn** [comment]
I think in the future if we actually so terribly need more wire layers for complexity, we could go the way of pipe layers (but, you know, not turbo garbage). Would let you have multiple separate wires in a small space without any accidental crossover.

**Comment by Shadowflame909** [comment]
The easier it is for engineering to do repairs the less reason the captain has to call the shuttle.

**Comment by actioninja** [comment]
I've barely tested this but all the basic logic seems to be working, and roundstart station is running fine.
Fuck it, let's roll.

**Comment by Jalleo** [comment]
Having multiple colours are suitable for a few reasons. All the maps do have seven powernets but are limited in scale.
The AI/Telecomms are seperate.
Solars and SM each have a powernet.
The rest of the station.

Something people could have done is feed all the maint and solar APCs and their power generation to engineering to centralise power production.

They didn't it would of required a lot more work.

**Comment by Razharas** [comment]
Wonderful pr
People who suggest side tooling for mapping cables are wrong on all imagibnable layers

**Comment by timothymtorres** [comment]
This PR is several years overdue.  Wire laying has been a clunky mess for several years now.  Glad to finally see that it’s been simplified to something akin to Minecraft redstone.  This should make the mechanic more intuitive for all players.

**Comment by SpinnerMaster** [comment]
This will be a good buff to the IQ of engineering. I know a lot of new players who wanted to try engineering but the non intuitive wire placement system that is currently in use turns a lot of people off of it entirely. This would also make life onboard the cryo pod station (that you can spawn at when you die) a lot more bearable.

**Comment by YakumoChen** [comment]
Will this remove my ability to make solars wire art aka turning solars wiring into a complete fucking mess

**Comment by zxaber** [comment]
Yes. You can do art with pipe cleaners but it won't be functional anymore.

File: code/modules/power/cable.dm
Comment (ninjanomnom) [inline code/modules/power/cable.dm]: switch on under_thing instead to make this all more readable

File: code/modules/power/cable.dm
Comment (ninjanomnom) [inline code/modules/power/cable.dm]: Put these defines outside the proc

File: code/modules/power/cable.dm
Comment (ninjanomnom) [inline code/modules/power/cable.dm]: Same here, put it outside the proc

File: code/modules/power/cable.dm
Comment (ninjanomnom) [inline code/modules/power/cable.dm]: Not needed

File: code/modules/power/cable.dm
Comment (ninjanomnom) [inline code/modules/power/cable.dm]: Why are you storing the path?

**Review by ninjanomnom — CHANGES_REQUESTED** [review]
I greatly hate this change but if I'm forced to accept it's happening at least I'm going to make sure we get the associated improved code from stripping out cable code

**Comment by SouDescolado** [comment]
I liked wiring the solars. Can engineering get a space suit for monkeys so they can wire the solars now?

File: code/modules/power/cable.dm
Comment (actioninja) [reply code/modules/power/cable.dm]: as far as I can tell, smoothing system wouldn't be able to do the conditional node sprites or conditionally not linking only in one direction.

**Comment by CRITAWAKETS** [comment]
Maybe add some cable blockers? Just doing crafting with a cable to make them so that you dont have to haul extra resources. Also, a way of doing diagonal wires would be nice. I love solar art.

File: code/modules/power/cable.dm
Comment (ninjanomnom) [reply code/modules/power/cable.dm]: The conditional node sprite can be applied as an overlay seperate from the other stuff here which is functionaly identical to smoothing

File: code/modules/power/cable.dm
Comment (ninjanomnom) [reply code/modules/power/cable.dm]: If you're not absolutely sure something is resolved leave it to the maintainer to resolve. Unresolved change requests do not harm your pr

Comment (actioninja) [reply code/modules/power/cable.dm]: I did resolve this and then it snuck back in via me being bad at merge conflict resolution.

Comment (actioninja) [reply code/modules/power/cable.dm]: https://github.com/tgstation/tgstation/commit/8a310cb5103cb8407f5d86a3564dd83980b6e273

File: code/modules/power/cable.dm
Comment (ninjanomnom) [reply code/modules/power/cable.dm]: Misread how this was getting used

**Comment by Tlaltecuhtli** [comment]
@actioninja i was on white ship with some emitters and a pacman, every time i moved the ship every fucking powernet connection broke, all emitters wouldnt detect the cable under it, pacman disconnects and also the straight cables would turn inactive and disconnect from the powernet

**Comment by WJohn** [comment]
Yeah, ship rotation seems to fuck with cables, it's something that needs fixing.

**Comment by Tlaltecuhtli** [comment]
wires can be placed on top of floortiles
wires are bomb proof

**Comment by WJohn** [comment]
Having tested both of those, that's not the case. You're confusing it with pipe cleaners, which CAN be placed on top of floors now (though they're a bit bug at the time of writing), and neither pipe cleaners nor wires are bomb proof, whether they're above or below the floor.

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/modules/power/cable.dm
Comment (actioninja) [reply code/modules/power/cable.dm]: it does. the continue is targeting the for on l43, skipping L59-63

Comment (actioninja) [reply code/modules/power/cable.dm]: didn't realize dm had loop labels, changed it to a clearer syntax

**Comment by 81Denton** [comment]
Noticed two things: The centcom.dmm map file refuses to load and solar cabling auto connects itself so that SMES input and output are connected to each other. Second one should be an easy fix since you just need to move the cables a little bit.

Capture

File: code/modules/power/cable.dm
Comment (ninjanomnom) [inline code/modules/power/cable.dm]:
```
if(locate(search_target) in TB)
  continue
```

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by actioninja** [comment]
wow NICE JOB wjohn what am i not paying you for

File: code/modules/power/cable.dm
Comment (actioninja) [reply code/modules/power/cable.dm]: I hate being like "mmmm nah lol" to a maintainer review, but I genuinely don't think smoothing ss would be a good fit for this or even work. The conditional nodes isn't a problem, but conditional linking is and on top of that I don't know if you could get the sprites to even work correctly for this, smoothing ss sprites need a special format that's not very well suited for anything that's not something like a table or a wall.

File: code/game/objects/items/RCL.dm
Comment (ninjanomnom) [reply code/game/objects/items/RCL.dm]: ^

**Comment by actioninja** [comment]
Alright I think this is ready for another round of testmerging.
There's a fairly big known issue with lagspikes when cutting intersection cables but I'm not entirely sure how to fix it.

File: code/modules/power/cable.dm
Comment (actioninja) [inline code/modules/power/cable.dm]: this is nasty and also makes some pretty nasty lagspikes. oranges said something about registering to ssobj but I'm not entirely sure what he meant by that

File: code/modules/power/power.dm
Comment (actioninja) [inline code/modules/power/power.dm]: doing all the application at the end like this may also be the source for lagspikes. It technically is only needed for very specific situations so I may make it a parameter as to where this is fired.

Comment (actioninja) [reply code/modules/power/power.dm]: ok yeah that was actually the problem.

**Comment by optimumtact** [comment]
Test merged on basil

**Comment by Tlaltecuhtli** [comment]
give cable stacks back to 30 or edit every place that uses cables like machinery, cable cuffs

**Comment by actioninja** [comment]
You can print cable for 10 metal and 5 glass unupgraded, and there's shitloads of coil roundstart all over the place.

**Comment by Tlaltecuhtli** [comment]
toolbelt holds only 15 so i cant make a cablecuff if i used 1 cable somewhere without having to use an extra slot out of my toolbelt, yellow toolboxes roundstart with 10 slots, autolathes are set to print 30 cable (i like it tho )
also bagil has bad ping http://take.ms/Z4ttt9 sybil is fine i think its cable's fault too

**Comment by adamkad1** [comment]
what does this mean to the future of wireart tho

**Comment by Tlaltecuhtli** [comment]
use spraycans

**Comment by actioninja** [comment]
I think this is ready. It should probably stay in testmerge for a bit longer to iron out bugs but it seems to be working fine.

**Comment by Tlaltecuhtli** [comment]
is pubby tcomms fixed?

**Comment by WJohn** [comment]
Yeah I fixed it. Completely missed that one earlier, haha.

**Comment by zxaber** [comment]
Tlaltecuhtli is right, if you're lowering cable stacks to 15 max, you should also lower how many it takes to make cable cuffs to 7 or 8.

Though I'd rather you didn't lower the stack at all; it's going to be annoying for Robotics to run out of cable every 15 minutes.

**Comment by adamkad1** [comment]
> Tlaltecuhtli is right, if you're lowering cable stacks to 15 max, you should also lower how many it takes to make cable cuffs to 7 or 8.
> 
> Though I'd rather you didn't lower the stack at all; it's going to be annoying for Robotics to run out of cable every 15 minutes.

good point

**Comment by WJohn** [comment]
The thing is that it now takes half as many cables as before to wire any floor, sometimes even less, but machines and the like haven't changed their wire costs to build.

The alternative is just to leave them at 30, but perhaps reduce the amount of cables found in electrical toolboxes.

**Comment by zxaber** [comment]
Leaving the max but lowering the starting amount sounds like a good compromise.

**Comment by kriskog** [comment]
I bet if someone checked, only a smaller fraction of the current starting cables are used for wiring. 5-6 stacks for all four solars, but thats probably pretty much it. While it might sound good in theory, balancing stack size and starting cables around that is a bit silly.

**Comment by zxaber** [comment]
By bulk, maybe, but robotics is using wires constantly. Every borg requires a wire, as does every mech, and as does doing wire repairs for both. Plus, since the amount you need to cuff someone is also 15, if you do anything at all with the wires you're SoL and have no way of non-lethally securing someone unless you're willing to sacrifice an extra slot for it.

**Comment by optimumtact** [comment]
I think that can be done in another pr, I don't want to hold this up becuase it will cause map conflicts

**Comment by MCterra10** [comment]
krSOPD3h
*sigh*
What have you brought upon this cursed land

**Comment by MCterra10** [comment]
I'd keep it out of my downstream but unfortunately I don't feel like dealing with all the merge conflicts

**Comment by adamkad1** [comment]
Fancy solars

**Comment by PolyGanon** [comment]
Encountered a duplication bug with this. I tried to replace a mouse-eaten cable in maintenance without any kind of gloves using the **single** piece of cable that was on the floor from it. When I was burnt, it dropped **two** wires instead.

I'd hazard a guess it's related to electrocution code of something along the lines of "If electrocuted when trying to complete a join, drop two wires".

For my two cents, I still prefer the old wire system.

**Comment by MCterra10** [comment]
This was merged way before being ready and with loads of bugs. While the change was made with good in mind, the execution was pretty shit and this was rushed without bugs being fixed. It looks ugly and is more clunky than the old system with literally everything except placing wires.

**Comment by optimumtact** [comment]
just remap ur map

**Comment by MCterra10** [comment]
> complain about orange man
> actually make changes to your repo and deal with the ensuing merge conflicts
Which would you choose?

**Comment by optimumtact** [comment]
complaining

**Comment by adamkad1** [comment]
so, maybe give rcd the ability to make cabled machines by default? the upgrade isnt too expensive but im sure research usually have bigger priorities than rcd upgrade tech

**Comment by optimumtact** [comment]
no?

**Comment by adamkad1** [comment]
> no?

pshhh fine, can we atleast have a higher stack size with pipe cleaners? its hard to do art with only 15 of em at once

**Comment by optimumtact** [comment]
it was changed in another PR

**Pull request merged**
